HELM-248: Document Filters (#614)
OpenNMS/opennms-helm: docs/modules/datasources/pages/flow_datasource.adoc

| withHost
| Filter for flows with the given host values.
Do not use with `topN`.

2+|*Transform functions*

| toBits
| Convert the values from bytes to bits.

| perSecond
| Convert the values to a per-second rate.

| negativeEgress
| Make egress values negative.

| negativeIngress
| Make ingress values negative.

| asTableSummary
| Generate a tabular summary instead of time series data.

| combineIngressEgress
| Combine ingress and egress values.

| onlyIngress
| Display only ingress traffic.

| onlyEgress
| Display only egress traffic.

| withGroupByInterval
| Changes the resolution of the returned datapoints grouping by the given time interval (`10s`, `5m`, `1h`, and so on).

| withPrefix
| Specifies a prefix for graph and summary labels.

| withSuffix
| Specifies a suffix for graph and summary labels.

| nanToZero
| Convert all `NaN` values to zero (`0`) so that reduce expressions (`sum`, `max`) can take them into consideration.
|===

[[ds-flow-template]]
== Template queries

=== locations()

The `locations()` query returns a list of available node locations.
You can use this inside an `exporterNodesWithFlows()` query to filter the number of nodes displayed by location.

> .Example expressions using `locations()`

Author: ```suggestion
.Example `exporterNodesWithFlows()` expressions
```

[source,]
----
exporterNodesWithFlows(location='Default')

> exporterNodesWithFlows(location='$location')
> ----

Author: ```suggestion
exporterNodesWithFlows(location='$location') <1>
----
<1> The `$location` variable is populated with a user-selected location.
```

> exporterNodesWithFlows(location='$location')
> ----

Reviewer: See note on performance datasource - technically these examples are not showing the use of `locations()`
```suggestion
exporterNodesWithFlows(location='$location') <1>
----
<1> Assumes a `$location` variable based on a `locations()` query.
```

Author: Could you provide an example of a `locations()` query, or is it one that accepts no arguments and simply returns a result?

Reviewer 2: locations() is meant to be used as a parameter 
When a user selects a location from the filter, the selected location can be used in directly in the query editor or in another template variable.

![Screenshot 2023-02-21 at 9 32 22 AM](https://user-images.githubusercontent.com/3297292/220375192-6b44cc58-02eb-4843-a2d3-5d8bcf72cdfd.png)

![Screenshot 2023-02-21 at 9 32 03 AM](https://user-images.githubusercontent.com/3297292/220375253-959911c8-35cd-4b14-8c52-7443fc54b192.png)

Reviewer: Yeah, this is a better example that shows using two variables to feed `locations()` into a `nodeFilter()`

Author: ```suggestion
nodeFilter(location='$location') <1>
----
<1> Assumes that the `$location` variable is populated with a value from a `locations()` query.
```

=== exporterNodesWithFlows(<filterQuery>)

The `exporterNodesWithFlows()` query returns a list of nodes that have exported flows data to OpenNMS.

> The value of the template value is the node's ID, and the display text is the node's label.

Author: ```suggestion
The template value is the node's ID, and the display text is the node's label.
```

| Argument  | Description

| nodeCriteria
| Criteria (node ID, or node foreign source and node foreign ID) of the node on which to enumerate values.

| interface
| Interface index of the SNMP interface to enumerate.

| from
| Beginning of the time range for which to return values, in epoch milliseconds.

| to
| End of the time range for which to return values, in epoch milliseconds.
|===

.Example expression using `dscpOnExporterNodeAndInterface()`
[source,]
----
dscpOnExporterNodeAndInterface($node, $interface, $__from, $__to) <1>
----
<1> Assumes that the `$node` variable is populated with a value from an `exporterNodesWithFlows()` query, and the `$interface` variable is populated with a value from the `interfacesOnExporterNodeWithFlows($node)` query.

=== applications(<limit>)

The `applications()` query returns a list of all available applications.
This query is intended for use inside the `withApplications()` filter function as a template variable, if necessary.

The query accepts an optional number of records, which ensures that only a limited number of records is retrieved.

NOTE: The application values retrieved come "as is" from the `flows/applications/enumerate` API.

> .Example expressions using `applications()`

Author: ```suggestion
.Example expressions
```

----
applications(1000)
withApplications($applications) <1>
----
<1> Assumes that the `$applications` variable is populated with a value from an `applications()` query.

=== conversations(<application>,<location>,<protocol>,<limit>)

The `conversations()` query returns a list of all available conversations.
This query is intended for use inside the `withConversation()` filter function as a template variable, if necessary.

NOTE: The conversations values retrieved come "as is" from the `flows/conversations/enumerate` API.

The query accepts the following optional arguments:

[options="autowidth"]
|===
| Argument  | Description   | Position  | Default Value

| application
| Application expression
| 1
| `.*`

| location
| Location expression
| 2
| `.*`

| protocol
| Protocol expression
| 3
| `.*`

| limit
| Limit for the number of records retrieved.
| Last
| 25
|===

> .Example expressions using `conversations()`

Author: ```suggestion
.Example expressions
```

withConversation($conversations) <5>
----
<1> Queries all conversations and returns a maximum of 10 records.
<2> Queries all conversations in the `app0` application and the default location, and returns a maximum of 10 records.
<3> Queries all conversations using the `http` protocol in the `app0` application and the default location.
Returns a maximum of 10 records.
<4> Queries all conversations using the `http` and `https` protocols in applications starting with `app`, in locations starting with `Def`.
Returns a maximum of 1,000 records.
<5> Assumes that the `$conversations` variable is populated with a value from a `conversations()` query.

=== hosts(<pattern>,<limit>)

The `hosts()` query returns a list of available hosts.
This query is intended for use inside the `withHost()` filter function as a template variable, if necessary.

NOTE: The host values retrieved come "as is" from the `flows/hosts/enumerate` API.

This query accepts the following optional arguments:

[options="autowidth"]
|===
| Argument  | Description   | Position  | Default Value

| pattern
| Host expression
| 1
| `.*`

| limit
| Limit for the number of records retrieved.
| Last
| 25
|===

> .Example expressions using `hosts()`

Author: ```suggestion
.Example expressions
```

| withConversation
| Filter for flows with the given conversation values.
Do not use with `topN`.

| withHost
| Filter for flows with the given host values.
Do not use with `topN`.

2+|*Transform functions*

| toBits
| Convert the values from bytes to bits.

| perSecond
| Convert the values to a per-second rate.

| negativeEgress
| Make egress values negative.

| negativeIngress
| Make ingress values negative.

| asTableSummary
| Generate a tabular summary instead of time series data.

| combineIngressEgress
| Combine ingress and egress values.

| onlyIngress
| Display only ingress traffic.

| onlyEgress
| Display only egress traffic.

| withGroupByInterval
| Changes the resolution of the returned datapoints grouping by the given time interval (`10s`, `5m`, `1h`, and so on).

| withPrefix
| Specifies a prefix for graph and summary labels.

| withSuffix
| Specifies a suffix for graph and summary labels.

| nanToZero
| Convert all `NaN` values to zero (`0`) so that reduce expressions (`sum`, `max`) can take them into consideration.
|===

[[ds-flow-template]]
== Template queries

=== locations()

The `locations()` query returns a list of available node locations.

> You can use this inside an `exporterNodesWithFlows()` query to filter the number of nodes displayed by location.

Reviewer: ```suggestion
You can create a variable with a `locations()` query and use that variable inside a `exporterNodesWithFlows()` query to filter the number of nodes displayed by location.
```

=== locations()

The `locations()` query returns a list of available node locations.

> You can create a variable with a `locations()` query and use that variable inside a `exporterNodesWithFlows()` query to filter the number of nodes displayed by location.

Reviewer 3: ```suggestion
You can create a variable with a `locations()` query and use that variable inside an `exporterNodesWithFlows()` query to filter the number of nodes displayed by location.
```
